### PR TITLE
Bugfix: Alert rules service should not be dependent on detectionlists, change dependency 

### PR DESCRIPTION
**Pull request by kiran-chaudhary**
### Description of Change ###

As `alertrules` service was dependent on `detectionlists` for `user-profile`, `rules.add_user` failed whenever a user existed but was not added to `detectionlists`.

EDIT:
   Added custom exceptions when
    -  user-id is invalid
    -  rule-id is invalid

The fix  changes dependency to `user service`. 

### Testing Procedure ###

`new_user = sdk.create_user(org_uid, "abc@abc.abc', "abc@abc.abc')`
`sdk.alerts.rules.add_user("rule-id", new_user['userId'])`

### PR Checklist ###
Did you remember to do the below?

- [x] Add unit tests to verify this change
- [NA ] Add an entry to CHANGELOG.md describing this change
- [ NA] Add docstrings for any new public parameters / methods / classes

### CODE REVIEW

**Comment by timabrmsn** [comment]
This doesn't actually resolve the issue, as the user still doesn't exist in the detectionlists microservice. This just allows the get_by_id call to succeed, but it wouldn't provide the needed `cloudUsernames` data we're making that call for:

```
user_details = self._user_profile_service.get_by_id(user_id)
user_aliases = user_details.data.get(u"cloudUsernames") or []
```

The `post` still fails. What we need to do is call `DetectionListUserService.create_if_not_exists()` to handle when a user isn't on any list and someone tries to add them to a rule.

**Comment by unparalleled-js** [comment]
I think the real issue is that we are not creating the profile when we should be automatically

**Comment by kiran-chaudhary** [comment]
Need confirmation on whether we want to create it, can a user be associated to alert-rules without being added to `detectionlists`?

**Comment by kiran-chaudhary** [comment]
> This doesn't actually resolve the issue, as the user still doesn't exist in the detectionlists microservice. This just allows the get_by_id call to succeed, but it wouldn't provide the needed `cloudUsernames` data we're making that call for:
> 
> ```
> user_details = self._user_profile_service.get_by_id(user_id)
> user_aliases = user_details.data.get(u"cloudUsernames") or []
> ```
> 
> The `post` still fails. What we need to do is call `DetectionListUserService.create_if_not_exists()` to handle when a user isn't on any list and someone tries to add them to a rule.

My guess, you might have passed id instead of UID while testing, the documentation of `user_profile.get_by_id` is incorrect, whereas the client interface has the correct documentation.

File: src/py42/services/alertrules.py [hunk 1]
```diff
@@ -41,6 +41,7 @@ def filetypemismatch(self):
 
     def add_user(self, rule_id, user_id):
         tenant_id = self._user_context.get_current_tenant_id()
+        self._user_profile_service.create_if_not_exists(user_id)
```
Comment (unparalleled-js) [inline src/py42/services/alertrules.py:44]: Missing tests?

**Comment by unparalleled-js** [comment]
Can we add a changelog entry:  `Fixes failure that occurred when trying to add a user to an alert rule who had never been on any detection list before.`

File: CHANGELOG.md [hunk 2]
```diff
@@ -11,6 +11,7 @@ how a consumer would use the library (e.g. adding unit tests, updating documenta
 ## Unreleased
 
 ### Fixed
+- Bug that occurrs when trying to add a user to an alert rule who had never been on any detection list before.
```
Comment (unparalleled-js) [inline CHANGELOG.md:14]: Should always be a newline under markdown headers

**Comment by timabrmsn** [comment]
> My guess, you might have passed id instead of UID while testing

I did exactly what your testing procedure says:

```
>>> new_user = sdk.users.create_user(944755074379520217, "abc2@example.com", "abc2@example.com")
>>> sdk.alerts.rules.add_user("3e4ccee2-f475-47c9-810d-8dc3e7d94222", new_user['userId'])
Traceback (most recent call last):
  File "<input>", line 1, in <module>
    sdk.alerts.rules.add_user("3e4ccee2-f475-47c9-810d-8dc3e7d94222", new_user['userId'])
  File "/Users/tim.abramson/PycharmProjects/py42/src/py42/clients/alertrules.py", line 49, in add_user
    return self._alert_rules_service.add_user(rule_id, user_id)
  File "/Users/tim.abramson/PycharmProjects/py42/src/py42/services/alertrules.py", line 54, in add_user
    return self._connection.post(uri, json=data)
  File "/Users/tim.abramson/PycharmProjects/py42/src/py42/services/_connection.py", line 145, in post
    return self.request(u"POST", url, data=data, **kwargs)
  File "/Users/tim.abramson/PycharmProjects/py42/src/py42/services/_connection.py", line 209, in request
    _handle_error(method, url, response)
  File "/Users/tim.abramson/PycharmProjects/py42/src/py42/services/_connection.py", line 274, in _handle_error
    raise_py42_error(ex)
  File "/Users/tim.abramson/PycharmProjects/py42/src/py42/exceptions.py", line 245, in raise_py42_error
    raise Py42BadRequestError(raised_error)
py42.exceptions.Py42BadRequestError: Failure in HTTP call 400 Client Error: Bad Request for url: https://fed-observer-default.prod.ffs.us2.code42.com/svc/api/v1/Rules/add-users
```

**Comment by kiran-chaudhary** [comment]
> > My guess, you might have passed id instead of UID while testing
> 
> I did exactly what your testing procedure says:
> 
> ```
> >>> new_user = sdk.users.create_user(944755074379520217, "abc2@example.com", "abc2@example.com")
> >>> sdk.alerts.rules.add_user("3e4ccee2-f475-47c9-810d-8dc3e7d94222", new_user['userId'])
> Traceback (most recent call last):
>   File "<input>", line 1, in <module>
>     sdk.alerts.rules.add_user("3e4ccee2-f475-47c9-810d-8dc3e7d94222", new_user['userId'])
>   File "/Users/tim.abramson/PycharmProjects/py42/src/py42/clients/alertrules.py", line 49, in add_user
>     return self._alert_rules_service.add_user(rule_id, user_id)
>   File "/Users/tim.abramson/PycharmProjects/py42/src/py42/services/alertrules.py", line 54, in add_user
>     return self._connection.post(uri, json=data)
>   File "/Users/tim.abramson/PycharmProjects/py42/src/py42/services/_connection.py", line 145, in post
>     return self.request(u"POST", url, data=data, **kwargs)
>   File "/Users/tim.abramson/PycharmProjects/py42/src/py42/services/_connection.py", line 209, in request
>     _handle_error(method, url, response)
>   File "/Users/tim.abramson/PycharmProjects/py42/src/py42/services/_connection.py", line 274, in _handle_error
>     raise_py42_error(ex)
>   File "/Users/tim.abramson/PycharmProjects/py42/src/py42/exceptions.py", line 245, in raise_py42_error
>     raise Py42BadRequestError(raised_error)
> py42.exceptions.Py42BadRequestError: Failure in HTTP call 400 Client Error: Bad Request for url: https://fed-observer-default.prod.ffs.us2.code42.com/svc/api/v1/Rules/add-users
> ```

Yes, it should be `new_user['userUid']` instead of `new_user['userId']`

File: tests/services/test_alert_rules.py [hunk 3]
```diff
@@ -27,6 +33,54 @@ def mock_detection_list_user_service(mocker, py42_response):
...
+    mocker, py42_response, http_error, mock_connection
+):
+    py42_response.text = MOCK_DETECTION_LIST_GET_RESPONSE
+    py42_response.data = json.loads(MOCK_DETECTION_LIST_GET_RESPONSE)
+    response = mocker.MagicMock(spec=Response)
+    response.status_code = 400
+    exception = mocker.MagicMock(spec=Py42BadRequestError)
+    exception.response = response
+    detection_list_user_service = mocker.MagicMock(spec=DetectionListUserService)
+    detection_list_user_service.get_by_id.return_value = py42_response
```
Comment (kiran-chaudhary) [inline tests/services/test_alert_rules.py:78]: Need to make this a side-effect as it gets called twice, first time it should fail with `BadRequest` error and then after the user is added by `create_user` it  should return valid response.

File: src/py42/services/alertrules.py (hunk 1 above)
Comment (kiran-chaudhary) [reply src/py42/services/alertrules.py]: added.

File: CHANGELOG.md (hunk 2 above)
Comment (kiran-chaudhary) [reply CHANGELOG.md]: changed.

File: src/py42/exceptions.py [hunk 4]
```diff
@@ -237,6 +237,14 @@ def __init__(self, exception, user_id, list_name):
         super(Py42NotFoundError, self).__init__(exception, message)
 
 
+class Py42InvalidRule(Py42NotFoundError):
```
Comment (unparalleled-js) [inline src/py42/exceptions.py:240]: class should end in `Error`

Comment (unparalleled-js) [reply src/py42/exceptions.py:240]: was `Py42InvalidRuleOperationError` intended for this? It seems the same

Comment (kiran-chaudhary) [reply src/py42/exceptions.py]: yea, we can merge them into one with proper messages being passed from respective module, at present, as per my understanding, the message in `py42InvalidRuleOperationError` doesn't fit this use case.

Added `Error`, renamed.

**Review by timabrmsn — APPROVED** [review]

**Pull request merged**